OpenRA.Mods.Common/AI/States/NavyStates.cs

```diff
@@ -0,0 +1,195 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.AI
+{
+	abstract class NavyStateBase : StateBase
+	{
+		protected virtual bool ShouldFlee(Squad owner)
+		{
+			return base.ShouldFlee(owner, enemies => !AttackOrFleeFuzzy.Default.CanAttack(owner.Units, enemies));
+		}
+
+		protected Actor FindClosestEnemy(Squad owner)
+		{
+			var first = owner.Units.First();
+
+			// Navy squad AI can exploit enemy naval production to find path, if any.
+			// (Way better than finding a nearest target which is likely to be on Ground)
+			var navalProductions = owner.World.ActorsHavingTrait<Building>().Where(a
+				=> owner.Bot.Info.BuildingCommonNames.NavalProduction.Contains(a.Info.Name)
+				&& a.AppearsHostileTo(first));
```

Comment:
This should do a `DomainIndex` check to prevent it from selecting targets in disconnected bodies of water.